### PR TITLE
fix(blockifier): merge state diff with squash

### PR DESCRIPTION
**Pull request by yoavGrs**

### CODE REVIEW

**Comment by yoavGrs** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2310?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2310** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2310?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2310?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2301** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2301?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2300** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2300?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2293** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2293?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2150** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2175** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2175?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2149** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2149?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2200** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2200?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 5 unresolved discussions (waiting on @yoavGrs and @Yoni-Starkware)

___
*[a discussion](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm8wBj8PdxlCjPR_k5:-OCm8wBj8PdxlCjPR_k6:b2kwqdb) (no related file):*
nice! much cleaner, and nice squashing algorithm :)
___
*[`crates/blockifier/src/state/cached_state.rs` line 62 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm7buW2gO8UdtFO7nA:-OCm7buW2gO8UdtFO7nB:b-aflr8e) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L62)):*
> ```XML
> 
>     pub fn to_state_cache(&mut self) -> StateResult<StateCache> {
>         self.update_initial_values_of_write_only_access()?;
> ```

what is this?



_Code quote:_
```XML
self.update_initial_values_of_write_only_access()?;
```
___
*[`crates/blockifier/src/state/cached_state.rs` line 395 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm7fvs7A5ev7pPel1L:-OCm7fvs7A5ev7pPel1M:bvhl5pn) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L395)):*
> ```XML
> 
> // Invariant: keys cannot be deleted from fields (only used internally by the cached state).
> #[derive(Clone, Debug, Default, PartialEq, Eq)]
> ```

I'd rather not ad this Clone if possible - the struct size is not O(1).
can you return a reference in to_state_cache?
or, change the signature to accept `mut self` instead of `&mut self`; `to_X` functions are consuming by convension



_Code quote:_
```XML
(Clone,
```
___
*[`crates/blockifier/src/state/cached_state.rs` line 417 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm89tDA0LAEW4ARwb6:-OCm89tDA0LAEW4ARwb7:bgeggsm) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L417)):*
> ```XML
>     /// is important.
>     pub fn merge(state_caches: Vec<Self>) -> Self {
>         let mut merged_state_cache = StateCache::default();
> ```

i think this is the terminology



_Suggestion:_
```XML
    /// Squashes the given state caches into a single one. Note that the order of the state caches
    /// is important.
    pub fn squash(state_caches: Vec<Self>) -> Self {
        let mut squashed_state_cache = StateCache::default();
```
___
*[`crates/blockifier/src/state/cached_state_test.rs` line 418 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm8kyIDoTMZ3CivfMO:-OCm8kyIDoTMZ3CivfMP:bb7lbs2) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state_test.rs#L418)):*
> ```XML
> #[case(true, vec![felt!("0x7"), felt!("0x0")], false)]
> #[case(false, vec![felt!("0x7"), felt!("0x1")], false)]
> #[case(false, vec![felt!("0x0"), felt!("0x8")], false)]
> ```

nice



_Code quote:_
```XML
#[case(false, vec![felt!("0x0"), felt!("0x8")], false)]
```

**Review by yoavGrs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @dorimedini-starkware and @Yoni-Starkware)

___
*[a discussion](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm8wBj8PdxlCjPR_k5:-OD5eJua3kbvMdsQEdHO:b32vhw) (no related file):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

nice! much cleaner, and nice squashing algorithm :)
</blockquote></details>

:)
___
*[`crates/blockifier/src/state/cached_state.rs` line 62 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm7buW2gO8UdtFO7nA:-OD5f2Dc7uZ2hYMe04BM:b-98cf58) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L62)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

what is this?
</blockquote></details>

It fills missing initial_reads, as done in `to_state_diff`.
___
*[`crates/blockifier/src/state/cached_state.rs` line 395 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm7fvs7A5ev7pPel1L:-OD5ysF50w3jyiTVdY0Q:bnua3wq) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L395)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I'd rather not ad this Clone if possible - the struct size is not O(1).
can you return a reference in to_state_cache?
or, change the signature to accept `mut self` instead of `&mut self`; `to_X` functions are consuming by convension
</blockquote></details>

I don't see an immediate solution to this.
In `account_transaction.rs`, we should save a "snapshot" of the validation state cache and update the state for the execution stage.

(Conceptually, only the `fn commit` needs a mutable base state, so it is possible to own a reference to the state cache up to this point.)
___
*[`crates/blockifier/src/state/cached_state.rs` line 417 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm89tDA0LAEW4ARwb6:-OD5fR9I6iWk31hzUTwI:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L417)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

i think this is the terminology
</blockquote></details>

Done.

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r5, 14 of 14 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 3 unresolved discussions (waiting on @yoavGrs and @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state.rs` line 395 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm7fvs7A5ev7pPel1L:-ODB1HOz3rZtajQMx46N:b2gwkvr) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L395)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

I don't see an immediate solution to this.
In `account_transaction.rs`, we should save a "snapshot" of the validation state cache and update the state for the execution stage.

(Conceptually, only the `fn commit` needs a mutable base state, so it is possible to own a reference to the state cache up to this point.)
</blockquote></details>

ok, I think it won't hit performance too much.
@Yoni-Starkware thoughts?
tl;dr extra `clone` of the state cache of a tx is now required
___
*[`crates/blockifier/src/state/cached_state.rs` line 412 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODB1YTi7TYM6Qqt2ILr:-ODB1YTi7TYM6Qqt2ILs:b-tbrto2) ([raw file](https://github.com/starkware-libs/sequencer/blob/1e93b82c203279078ad3124d1fabb15c5c2aae47/crates/blockifier/src/state/cached_state.rs#L412)):*
> ```XML
> 
>     /// Squashes the given state caches into a single one and returns the state diff. Note that the
>     /// order of the state caches is important.
> ```

please document the `comprehensive_state_diff` parameter



_Code quote:_
```XML
    /// Squashes the given state caches into a single one and returns the state diff. Note that the
    /// order of the state caches is important.
```
___
*[`crates/blockifier/src/state/cached_state_test.rs` line 459 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODB2CeW6KxjhFkEJVUz:-ODB2CeW6KxjhFkEJVV-:bnyg8eo) ([raw file](https://github.com/starkware-libs/sequencer/blob/1e93b82c203279078ad3124d1fabb15c5c2aae47/crates/blockifier/src/state/cached_state_test.rs#L459)):*
> ```XML
>         // allocated keys feature was inserted.
>         assert_ne!(merged_changes.allocated_keys.is_empty(), charged);
>     }
> ```

why is this inequality only expected if you squash?
if you are not squashing the booleans should be equal? if so please assert this



_Code quote:_
```XML
    if comprehensive_state_diff {
        // The comprehensive_state_diff is needed for backward compatibility of versions before the
        // allocated keys feature was inserted.
        assert_ne!(merged_changes.allocated_keys.is_empty(), charged);
    }
```

**Review by yoavGrs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 16 of 17 files reviewed, 3 unresolved discussions (waiting on @dorimedini-starkware and @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state_test.rs` line 459 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODB2CeW6KxjhFkEJVUz:-ODBBBYb-djgTpmE2hsp:bl40ino) ([raw file](https://github.com/starkware-libs/sequencer/blob/1e93b82c203279078ad3124d1fabb15c5c2aae47/crates/blockifier/src/state/cached_state_test.rs#L459)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

why is this inequality only expected if you squash?
if you are not squashing the booleans should be equal? if so please assert this
</blockquote></details>

Discussed.

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 3 unresolved discussions (waiting on @yoavGrs and @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state_test.rs` line 459 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODB2CeW6KxjhFkEJVUz:-ODBcCTy48kJ95tkK24o:b-p4b486) ([raw file](https://github.com/starkware-libs/sequencer/blob/1e93b82c203279078ad3124d1fabb15c5c2aae47/crates/blockifier/src/state/cached_state_test.rs#L459)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

Discussed.
</blockquote></details>

we said you'll add a unit test for backwards-compatibility, right? can you do it in this PR?

**Review by yoavGrs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 3 unresolved discussions (waiting on @dorimedini-starkware and @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state_test.rs` line 459 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODB2CeW6KxjhFkEJVUz:-ODC2dtSCu3t6YvEEfxE:bpf8kiv) ([raw file](https://github.com/starkware-libs/sequencer/blob/1e93b82c203279078ad3124d1fabb15c5c2aae47/crates/blockifier/src/state/cached_state_test.rs#L459)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

we said you'll add a unit test for backwards-compatibility, right? can you do it in this PR?
</blockquote></details>

This test checks backward compatibility, but only for the storage. Look at `expected_storage_diff`.

**Review by dorimedini-starkware — COMMENTED** [review]
> Reviewed 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 1 unresolved discussion (waiting on @Yoni-Starkware)

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 3 unresolved discussions (waiting on @dorimedini-starkware and @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 395 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-OCm7fvs7A5ev7pPel1L:-ODKb4Ex6vbq5vpZOIhH:b6b0o29) ([raw file](https://github.com/starkware-libs/sequencer/blob/8640df0ac14941b053a0fc7ffb58951d708dd9ab/crates/blockifier/src/state/cached_state.rs#L395)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

ok, I think it won't hit performance too much.
@Yoni-Starkware thoughts?
tl;dr extra `clone` of the state cache of a tx is now required
</blockquote></details>

If you define `let mut execution_state = CachedState::new(MutRefState::new(validate_state));`, you could access the cache of validate after execution. But it might be too ugly, so I'm okay with this clone.
___
*[`crates/blockifier/src/state/cached_state.rs` line 414 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODKiTzC9ZwY1Exn3J0h:-ODKiTzC9ZwY1Exn3J0i:botpd06) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9454fcbb37b37c0ca4cfd0f83321d64a9d3fd63/crates/blockifier/src/state/cached_state.rs#L414)):*
> ```XML
>     /// If 'comprehensive_state_diff' is false, opposite updates may not be canceled out. Used for
>     /// backward compatibility.
>     pub fn squashed_state_diff(
> ```

I think it's better to have a separate function that only returns a squashed StateCache (in your smart ordering)
___
*[`crates/blockifier/src/state/cached_state.rs` line 415 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODKimAc9XHaAAJ7fbbp:-ODKimAc9XHaAAJ7fbbq:b-qr7qc) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9454fcbb37b37c0ca4cfd0f83321d64a9d3fd63/crates/blockifier/src/state/cached_state.rs#L415)):*
> ```XML
>     /// backward compatibility.
>     pub fn squashed_state_diff(
>         state_caches: Vec<Self>,
> ```

Can you at least save the clone of `execute`? it is much heavier



_Suggestion:_
```XML
state_caches: Vec<&Self>,
```

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 4 unresolved discussions (waiting on @dorimedini-starkware and @yoavGrs)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 601 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODKj53D81yKr5UiMpI3:-ODKj53D81yKr5UiMpI4:b-mj7fq7) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9454fcbb37b37c0ca4cfd0f83321d64a9d3fd63/crates/blockifier/src/transaction/account_transaction.rs#L601)):*
> ```XML
>             ),
>             execution_steps_consumed,
>         );
> ```

@dorimedini-starkware any fancy way to save this iteration over the cache if not needed? (I.e., if the tx is not reverted)



_Suggestion:_
```XML
        // Save the state changes resulting from running `validate_tx`, to be used later for
        // resource and fee calculation.
        let validate_state_cache = state.to_state_cache()?;

        // Create copies of state and validate_resources for the execution.
        // Both will be rolled back if the execution is reverted or committed upon success.
        let mut execution_state = TransactionalState::create_transactional(state);

        let execution_result =
            self.run_execute(&mut execution_state, &mut execution_context, remaining_gas);

        // Pre-compute cost in case of revert.
        // TODO(tzahi): add reverted_l2_gas to the receipt.
        let execution_steps_consumed =
            n_allotted_execution_steps - execution_context.n_remaining_steps();
        let revert_receipt = TransactionReceipt::from_account_tx(
            self,
            &tx_context,
            &validate_state_cache.to_state_diff(),
            CallInfo::summarize_many(
                validate_call_info.iter(),
                &tx_context.block_context.versioned_constants,
            ),
            execution_steps_consumed,
        );
```

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 14 of 17 files reviewed, 5 unresolved discussions (waiting on @dorimedini-starkware and @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 69 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODL5SYf6GQ4ILACI6h7:-ODL5SYf6GQ4ILACI6h8:b-vy4myw) ([raw file](https://github.com/starkware-libs/sequencer/blob/3a7681ab5dd2e536ff33ef8a56f15fdc639c6828/crates/blockifier/src/state/cached_state.rs#L69)):*
>    /// It is not guaranteed that every write has a corresponding initial read.
Why are we okay with this?



_Code quote:_
```XML
    pub fn to_state_cache(&mut self) -> StateResult<StateCache> {
        self.update_initial_values_of_write_only_access()?;
        Ok(self.cache.borrow().clone())
    }

    /// It is not guaranteed that every write has a corresponding initial read.
    pub fn raw_state_cache(&self) -> Ref<'_, StateCache> {
        self.cache.borrow()
    }
```

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 14 of 17 files reviewed, 4 unresolved discussions (waiting on @dorimedini-starkware and @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 414 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODKiTzC9ZwY1Exn3J0h:-ODL5kAS09n42vhfOQ24:bc8gmc2) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9454fcbb37b37c0ca4cfd0f83321d64a9d3fd63/crates/blockifier/src/state/cached_state.rs#L414)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

Done.
(If I understood correctly what you asked me)
</blockquote></details>

Almost - just move the `.to_state_diff` outside (and rename `squashed_state_diff()` -> `squash_state_caches`

**Review by Yoni-Starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 14 of 17 files reviewed, 5 unresolved discussions (waiting on @dorimedini-starkware and @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 69 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODL6cmZEPWU3e2tI6c1:-ODL6cmZEPWU3e2tI6c2:b-4of6zf) ([raw file](https://github.com/starkware-libs/sequencer/blob/3a7681ab5dd2e536ff33ef8a56f15fdc639c6828/crates/blockifier/src/state/cached_state.rs#L69)):*
> ```XML
>     pub fn raw_state_cache(&self) -> Ref<'_, StateCache> {
>         self.cache.borrow()
>     }
> ```

Oh, I see that you moved it outside. If there is not particular reason, I think it's better to have a single function like this. You can clone the result outside for validate.



_Suggestion:_
```XML
    pub fn borrow_updated_state_cache(&mut self) -> StateResult<Ref<'_, StateCache>> {
        self.update_initial_values_of_write_only_access()?;
        Ok(self.cache.borrow())
    }
```

**Review by yoavGrs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 14 of 17 files reviewed, 5 unresolved discussions (waiting on @dorimedini-starkware and @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state.rs` line 414 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODKiTzC9ZwY1Exn3J0h:-ODLErhOClYsuCiF3r8i:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9454fcbb37b37c0ca4cfd0f83321d64a9d3fd63/crates/blockifier/src/state/cached_state.rs#L414)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Almost - just move the `.to_state_diff` outside (and rename `squashed_state_diff()` -> `squash_state_caches`
</blockquote></details>

Done.
___
*[`crates/blockifier/src/state/cached_state.rs` line 69 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODL5SYf6GQ4ILACI6h7:-ODLD2pz7MPOkhyDHVhj:bss4a2b) ([raw file](https://github.com/starkware-libs/sequencer/blob/3a7681ab5dd2e536ff33ef8a56f15fdc639c6828/crates/blockifier/src/state/cached_state.rs#L69)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

>    /// It is not guaranteed that every write has a corresponding initial read.
Why are we okay with this?
</blockquote></details>

Removed.
___
*[`crates/blockifier/src/state/cached_state.rs` line 69 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODL6cmZEPWU3e2tI6c1:-ODLD3xODinr4BvRaJvU:bo5nztk) ([raw file](https://github.com/starkware-libs/sequencer/blob/3a7681ab5dd2e536ff33ef8a56f15fdc639c6828/crates/blockifier/src/state/cached_state.rs#L69)):*
<details><summary><i>Previously, Yoni-Starkware (Yoni) wrote…</i></summary><blockquote>

Oh, I see that you moved it outside. If there is not particular reason, I think it's better to have a single function like this. You can clone the result outside for validate.
</blockquote></details>

Nice. Done.

**Review by yoavGrs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: 14 of 17 files reviewed, 5 unresolved discussions (waiting on @dorimedini-starkware and @Yoni-Starkware)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 601 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODKj53D81yKr5UiMpI3:-ODLGRLA7bGGfd_00pzJ:b-mevoj5) ([raw file](https://github.com/starkware-libs/sequencer/blob/c9454fcbb37b37c0ca4cfd0f83321d64a9d3fd63/crates/blockifier/src/transaction/account_transaction.rs#L601)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

If we don't return one big `match`, it's possible.
</blockquote></details>

https://reviewable.io/reviews/starkware-libs/sequencer/2471

**Review by dorimedini-starkware — APPROVED** [review]
> Reviewed 3 of 3 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 4 unresolved discussions (waiting on @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state_test.rs` line 447 at r12](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODLHllQ0tG5XwShc_fF:-ODLHllQ0tG5XwShc_fG:bwvhp11) ([raw file](https://github.com/starkware-libs/sequencer/blob/dc573fee4b1f18b2b542fe5aa2bee381582314de/crates/blockifier/src/state/cached_state_test.rs#L447)):*
> ```XML
>     }
> 
>     let merged_changes = StateCache::squash_state_diff_backward_compatible(
> ```

I don't think it's necessary to append `_backward_compatible` to the name; you have the (documented) boolean parameter
non-blocking



_Code quote:_
```XML
squash_state_diff_backward_compatible
```

**Review by yoavGrs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 4 unresolved discussions (waiting on @Yoni-Starkware)

___
*[`crates/blockifier/src/state/cached_state_test.rs` line 447 at r12](https://reviewable.io/reviews/starkware-libs/sequencer/2310#-ODLHllQ0tG5XwShc_fF:-ODLRpB2E7XvL0RLp36L:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/dc573fee4b1f18b2b542fe5aa2bee381582314de/crates/blockifier/src/state/cached_state_test.rs#L447)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I don't think it's necessary to append `_backward_compatible` to the name; you have the (documented) boolean parameter
non-blocking
</blockquote></details>

Done.

**Review by dorimedini-starkware — APPROVED** [review]
> Reviewed 3 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: all files reviewed, 4 unresolved discussions (waiting on @Yoni-Starkware)

**Review by Yoni-Starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @yoavGrs)

**Review by Yoni-Starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2310)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @yoavGrs)

**Comment by yoavGrs** [comment]
### Merge activity

* **Dec 5, 7:08 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/starkware-libs/sequencer/2310).
